sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala

```diff
@@ -388,6 +394,9 @@ private[hive] object HiveTableUtil {
 private[hive] object DeserializerLock
 
 private[hive] object HadoopTableReader extends HiveInspectors with Logging {
+
+  val avroTableProperties = AvroTableProperties.values().map(_.getPropName()).toSet
```

Comment:
SPARK-26836 is filed for 2.3.1 ~ 3.2.0 and this patch depends on `org.apache.hadoop.hive.serde2.avro.AvroSerdeUtils.AvroTableProperties`. I'm wondering if this patch is applicable with old Spark branches with old Hive libraries. For example,
- Apache Spark master : 2.3.8
- Apache Spark branch-3.1: 2.3.7
- Apache Spark branch-3.0: 2.3.7 and 1.2.1.spark2
- Apache Spark branch-2.4: 1.2.1.spark2

What is the minimum version of Hive library to use this? Is there a known bug for this `AvroTableProperties` classes?

Reply:
According to Hive communtiy, Apache Spark cannot access the following two table properties because it's in Apache Hive 4 via HIVE-21218 . I'm wondering if this can cause some issues in the future.
```
AVRO_SERDE_TYPE("avro.serde.type"),
AVRO_SERDE_SKIP_BYTES("avro.serde.skip.bytes"),
```

Reply:
The `AvroTableProperties` enum is used for avoiding the hardcoding of those property names which related to Avro and thus where the table properties has higher priority when a value for that property is chosen.

So this PR does NOT set any of those properties if the user hasn't specified one by using the `SERDEPROPERTIES` on the table or on the partition here we just change which one is wining before passing the values further to the SerDe initialization.

The `avro.schema.literal` on its own is not enough for example if the user used different `avro.schema.url` on the table than on the partition we will have the same issue. This can be true for all the properties it is good if they are coming from the consistent source: either from the table or from the partition. The advantage of using `AvroTableProperties` that we should not investigate all the consistency between these properties but just take them all from the same source and when a new one is introduced in the future which matters for the serde we will be still safe and this part does not generate a new bug silently (as we cannot have test for not existing future serde properties).

What do you think?

Reply:
@dongjoon-hyun I can offer a compromised solution where `AvroTableProperties` is not used at all. Instead of that the Avro property names will be detected with a `startsWith("avro.")` check.

 Is it ok with you?

Reply:
It seems that the examples of `avro.serde.type` and `avro.serde.skip.bytes` provide good reason why we shouldn't pick up `AvroTableProperties.*` or `startsWith("avro.")`. These properties indicate something about the specific files, e.g. it's possible you could have some partitions serialized with the Confluence Avro serializer (thus requiring the `avro.serde` properties) and some partitions serialized with normal Avro behavior. Schemas seem to be a special case.

Looking at the full list of properties in Hive 2.3:
```
    SCHEMA_LITERAL("avro.schema.literal"),
    SCHEMA_URL("avro.schema.url"),
    SCHEMA_NAMESPACE("avro.schema.namespace"),
    SCHEMA_NAME("avro.schema.name"),
    SCHEMA_DOC("avro.schema.doc"),
    AVRO_SERDE_SCHEMA("avro.serde.schema"),
    SCHEMA_RETRIEVER("avro.schema.retriever");
```
It seems that the first 5 we would definitely want at the table level, and probably the last one (?). I don't understand `avro.serde.schema` and how it relates to the others. But it might be reasonable to say that we should take all `startsWith("avro.schema.")`


Separately from the question of _which_ properties to pick up, there is the question of whether to hard-code the strings or use the enum values. @dongjoon-hyun can you explain more fully why you don't want to depend on `AvroTableProperties`? We already have another reference to it in the Spark codebase [within HiveShim](https://github.com/apache/spark/blob/6a194f197819e2970f619ed92321e0a0b716bf37/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala#L92). I don't understand the argument that some older branches may not have this class available -- if an older branch has a Hive dependency that is too old and we need to cherry pick this change there, then we can hard-code it in the cherry pick, but I don't see a reason why we can't use the enum in newer versions.

Reply:
@xkrogen thanks for the correct pattern: `startsWith("avro.schema.")`!

I will use that one.

Reply:
@attilapiros would you mind looking into what `avro.serde.schema` is used for? Understanding that better would provide more confidence (for me, at least) that `avro.schema` is the correct prefix.

Reply:
Sorry I read this message after my git push, but sure I will check that.

Reply:
Based on new discussion, sounds like we can leave that as a follow-on for later.

Reply:
Yes, but thanks for all your inputs.